File: apps/mobile/lib/domain/entities/saved_location.dart
/// Mirrors packages/contracts/src/savedLocation.ts#SavedLocation.
class SavedLocation {
  final String id;
  final String? entityId;
  final String label;
  final String? note;
  final String? sourceUrl;
  final double? latitude;
  final double? longitude;
  final String createdAt;

  const SavedLocation({
    required this.id,
    required this.entityId,
    required this.label,
    required this.note,
    required this.sourceUrl,
    required this.latitude,
    required this.longitude,
    required this.createdAt,
  });

  factory SavedLocation.fromJson(Map<String, dynamic> json) {
    return SavedLocation(
      id: json['id'] as String,
      entityId: json['entityId'] as String?,
      label: json['label'] as String,
      note: json['note'] as String?,
      sourceUrl: json['sourceUrl'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      createdAt: json['createdAt'] as String,
    );
  }

  bool get hasCoordinates => latitude != null && longitude != null;
}

File: packages/contracts/src/savedLocation.ts
import { z } from "zod";

/**
 * Section 48 (Favorites/Saved Locations). A saved location either points at
 * a known Entity (favorited from search/details) or is a freeform place the
 * user added by pasting a Google Maps link/address - see
 * packages/domain/src/googleMaps for how sourceUrl gets parsed into
 * name/latitude/longitude server-side.
 */
export const SavedLocation = z.object({
  id: z.string(),
  entityId: z.string().nullable(),
  label: z.string(),
  note: z.string().nullable(),
  sourceUrl: z.string().nullable(),
  latitude: z.number().nullable(),
  longitude: z.number().nullable(),
  createdAt: z.string().datetime(),
});
export type SavedLocation = z.infer<typeof SavedLocation>;

export const AddSavedLocationInput = z
  .object({
    entityId: z.string().optional(),
    label: z.string().min(1).max(140).optional(),
    note: z.string().max(1000).optional(),
    sourceUrl: z.string().url().max(2000).optional(),
  })
  .refine((v) => Boolean(v.entityId || v.label || v.sourceUrl), {
    message: "Provide at least one of entityId, label, or sourceUrl.",
  });
export type AddSavedLocationInput = z.infer<typeof AddSavedLocationInput>;
